src/scanner.c

```diff
@@ -1,127 +1,170 @@
+		__set_char_tok(source, AMPERSAND);
+	else if (matches_pipe(*source))
+		__set_char_tok(source, PIPE);
+	else if (matches_at(*source))
+		__set_char_tok(source, AT);
+	else if (matches_quote(*source))
+		__set_char_tok(source, QUOTE);
+	else if (matches_apostrophe(*source))
+		__set_char_tok(source, APOSTROPHE);
+	else if (matches_capture(*source))
+		__set_nlen_tok(source, 7, CAPTURE);
+	else if (matches_maybe(*source))
+		__set_nlen_tok(source, 5, MAYBE);
+	else if (matches_int(*source))
+		__set_nlen_tok(source, 3, INT);
+	else if (matches_bool(*source))
+		__set_nlen_tok(source, 4, BOOL);
+	else if (matches_char(*source))
+		__set_nlen_tok(source, 4, CHAR);
+	else if (matches_double(*source))
+		__set_nlen_tok(source, 6, DOUBLE);
+	else if (matches_long(*source))
+		__set_nlen_tok(source, 4, LONG);
+	else if (matches_unsigned(*source))
+		__set_nlen_tok(source, 8, UNSIGNED);
+	else if (matches_question(*source))
+		__set_char_tok(source, QUESTION);
+	else if (matches_and(*source))
+		__set_nlen_tok(source, 3, AND);
+	else if (matches_or(*source))
+		__set_nlen_tok(source, 2, OR);
+	else if (matches_just(*source))
+		__set_nlen_tok(source, 4, JUST);
+	else if (matches_notoken(*source) || (**source) == '\0')
+		__set_char_tok(source, NOTOKEN);
+	else if ((imgLen = matches_integer(*source, &tmp))) {
+		token->image = tmp;
+		token->tok = INTEGER;
+		*source += imgLen;
+	} else if ((imgLen = matches_float(*source, &tmp)))
+		__set_nlen_tok(source, imgLen, FLOAT);
+	else {
+		fprintf(stderr, "error: invalid token \"%s\"\n", *source);
+		return NULL;
+	} return token;
 }
 
 struct token_list * scan(const char * filename)
 {
-		char * source = calloc(129, sizeof(char));
-		int fd = -1; size_t bytes_read = 0;
-		size_t source_size = 0;
-
-		if ((fd = open(filename, O_RDONLY, 0600)) == -1) {
-				perror("open");
-				fprintf(stderr, "filename: \"%s\"\n", filename);
-				exit(1);
-		}
-
-		/* read the file */
-		for (; bytes_read = read(fd, source + source_size, 128);) {
-				source = realloc(source, source_size += bytes_read);
-		} source[source_size++] = '\0';
-
-		/* if source is empty, return NULL */
-		if (!source_size) return NULL;
-
-		/* create our list */
-		struct token_list * tokens = malloc(sizeof(*tokens));
-		tokens->head = NULL;
-
-		/* iterate over our source, gathering tokens */
-		struct token * tok = NULL; char * iter = source;
-		for (; *iter && (tok = get_next_token((const char **) &iter));) {
-				if (tok->tok != NOTOKEN) add_token(tokens, tok);
-				__free_token(tok);
-		}
-
-		free(source);
-		return tokens;
+	char * source = calloc(129, sizeof(char));
+	int fd = -1; size_t bytes_read = 0;
+	size_t source_size = 0;
+
+	if ((fd = open(filename, O_RDONLY, 0600)) == -1) {
+		perror("open");
+		fprintf(stderr, "filename: \"%s\"\n", filename);
+		exit(1);
+	}
+
+	/* read the file */
+	for (; bytes_read = read(fd, source + source_size, 128); )
+		source = realloc(source, source_size += bytes_read);
+	source[source_size++] = '\0';
```

Comment:
After the second realloc, the size of source will lose the additional 129 bytes intended for buffering, because it does not realloc to be `129 + source_size += bytes_read` the way you were thinking when you wrote this. So, after the first reallocation, for example, if it read only 60 bytes in a very short file, it will reallocate to 0+60.
The loop exits since the next `read` call will return 0 bytes read, but this assumption that 1 byte of memory always exists past the end of the allocated buffer is an invalid assumption in that case, which would cause this to situationally segfault.

Reply:
> The loop exits since the next read call will return 0 bytes read, but this assumption that 1 byte of memory always exists past the end of the allocated buffer is an invalid assumption in that case, which would cause this to situationally segfault.

I don't understand what you're saying here. Do you have a fix for it?